program/src/processor.rs

```diff
@@ -640,44 +695,108 @@ pub fn process_withdraw_inactive_stake(
             account_balance
         );
 
-        let excess_removed_here = withdraw_excess_inactive_sol(
-            &accounts,
-            &clock,
-            &rent,
-            &stake_history,
-            stake_account,
+        let excess_removed_here = withdraw_inactive_sol(
+            &WithdrawExcessOpts {
+                accounts: &accounts,
+                clock: &clock,
+                rent: &rent,
+                stake_history: &stake_history,
+                stake_account,
+            },
             seed,
             lido.stake_authority_bump_seed,
+            StakeType::Stake,
         )?;
 
         excess_removed = (excess_removed + excess_removed_here)?;
         observed_total = (observed_total + account_balance)?;
     }
 
-    if observed_total < validator.entry.stake_accounts_balance {
+    if observed_total < validator.entry.effective_stake_balance() {
         // Solana has no slashing at the time of writing, and only Solido can
         // withdraw from these accounts, so we should not observe a decrease in
         // balance.
         msg!(
             "Observed balance of {} is less than tracked balance of {}.",
             observed_total,
-            validator.entry.stake_accounts_balance
+            validator.entry.effective_stake_balance()
         );
         msg!("This should not happen, aborting ...");
         return Err(LidoError::ValidatorBalanceDecreased.into());
     }
 
     // We tracked in `stake_accounts_balance` what we put in there ourselves, so
     // the excess is a donation by some joker.
-    let donation = (observed_total - validator.entry.stake_accounts_balance)
+    let donation = (observed_total - validator.entry.effective_stake_balance())
         .expect("Does not underflow because observed_total >= stake_accounts_balance.");
     msg!("{} in donations observed.", donation);
 
     // Store the new total. If we withdrew any inactive stake back to the
     // reserve, that is now no longer part of the stake accounts, so subtract
     // that.
-    validator.entry.stake_accounts_balance =
-        (observed_total - excess_removed).expect("Does not underflow, because excess <= total.");
+    validator.entry.stake_accounts_balance = (validator.entry.stake_accounts_balance
+        - excess_removed)
+        .expect("Does not underflow, because excess <= total.");
```

Comment:
This needs to move to the end of the function; we don’t know the new total yet before looking at the unstake accounts.

Reply:
Yes, that’s the new total balance, and the new unstake balance is the `observed_unstake_total - unstake_removed`.